README.md
mod_authnz_persona is a module for Apache 2.0 or later that allows you to
quickly add Persona authentication to a site hosted with Apache.

# Installation

First, install the dependencies:

* apache 2.2 or later
* libcurl 7.10.8 or later (for remote verification)
* yajl 2.0 or later (for JSON parsing)

On Red Hat-derivative distributions, this might help:

```
yum install httpd httpd-devel curl-devel cmake
wget http://fedora.mirror.nexicom.net/linux/development/rawhide/source/SRPMS/y/yajl-2.0.4-3.fc20.src.rpm
rpmbuild --rebuild yajl-*.src.rpm
sudo yum install ~/rpmbuild/RPMS/`uname -i`/yajl-*.rpm
```

(CMake is a build-time yajl dependency.)

Then, clone the source and build:

```
git clone https://github.com/gozer-mozilla/mod_authn_persona.git
cd mod_authnz_persona
./buildconf
./configure --with-apxs=/usr/sbin/apxs
sudo make install
```

(this assumes apxs is behaving properly on your system; use --with-apxs
to point to your apxs or apxs2 binary of choice)

# Configuration

Configure the module:

    LoadModule authnz_persona_module modules/mod_authnz_persona.so
    
    #AuthPersonaServerSecret XXXSomeVerySecretUniqueString
    #AuthPersonaCookieName   Persona
    #AuthPersonaVerifierURL  https://verifier.login.anosrep.org/verify
    #AuthPersonaLoginURL     /login.shtml
    
    <Location />
       AuthType Persona
       Require valid-user
       # Or, require users with host/IdP example.com:
       # Require persona-idp example.com
       # Or, require specific users (requires mod_authz_user)
       # Require user user@example.com
    </Location>

This will cause the module to require Persona authentication for all
requests to the server.

Dependencies
============

* apache 2.0.x or 2.2.x (mostly tested with 2.2 so far)
* libcurl 7.10.8 or later
* json-c
* OpenSSL

# Features

* **zero configuration** - The module is designed with reasonable defaults, so
  you can simply drop it in.
* **automatic re-auth** - The module is designed to use session cookies and
  automatically re-authenticate.

# How it works

The module works by intercepting requests bound for protected resources, and
checking for the presence of a session cookie.

If the cookie is not found, the user agent is served an HTML document that
presents a Persona login page.

Upon successful authentication with Persona, this page will send a request to
the server with a Persona assertion in an HTTP header. The module, upon
detecting no cookie is present, will look for this header, validate the
assertion, and set a short session cookie.

The authentication page will then reload the desired resource.

# Further configuration settings

* `AuthPersonaServerSecret`: (default: generated)
  A secret that will be used to sign cookies. Must be set in a server or
  VirtualHost context. If not provided, upon server start a secret will be
  generated automatically. Given re-authentication is automatic, it is only
  required to set a cookie secret if your application is hosted on multiple
  load-balanced Apache servers.

* `AuthPersonaCookieName`: (default: Persona)

  The name for the Persona Session cookie

* `AuthPersonaCookieDomain`: (default: None)

  The domain that the cookie is valid for, so for instance '.domain.com'

* `AuthPersonaCookiePath`: (default: /)

  The url path the cookie is being set for

* `AuthPersonaCookieSecure`: (default: Off)

  If the cookie should be limited to SSL connections or not
  
  Sending this cookie via non-SSL connections is dangereous, as stealing this
  cookie steals your identity.

* `AuthPersonaCookieDuration`: (default: 1 day)

  The lifetime of the session itself, after that, the user will be re-authed

* `AuthPersonaAuthoritative`: (default: Off)

  Wherever the module will let other modules try their luck at authenticating users

* `AuthPersonaVerifierURL`: (default: https://verifier.login.persona.org/verify)

  This module uses a web service to verify recieved assertions, this is the URL
  of the service providing that service. It needs to be trusted

* `AuthPersonaLoginURL`: (default: /login.shtml)

  The location of the login page, it's never accessed directly. But it needs to be
  accessible by unauthenticated users.
* `AuthPersonaLogoutURL` (default: unset)

  The location where logout happens, accessing it triggers a session logout, but the
  location is responsible for the Persona JavaScript logout itself

* `AuthPersonaLogoutReturnTo` (default: /)

  When the logout page is accessed, where to redirect the now logged-out user to

* `AuthPersonaFakeBasicAuth`: (default: Off)

  When set, this will set a fake Authorization: header with the Persona userid and
  a fake password

* `Require persona-idp`:
  Only allow users with email addresses backed by the given Identity Provider.
  Note that this will often, but not necessarily, be the host part of the
  verified email address, in the case of email addresses backed by a secondary
  IdP (like the fallback IdP or a bridging IdP).


ChangeLog
0.5
 * Add AuthPersonaCookiePath directive to configure per-location cookie path
0.4 26-Nov-2013
 * Rename persona header to Persona-Assertion as per RFC 6648
 * Make module non-authoritative by default
 * Support for httpd-2.4
 * Security Review 2013-11-01 Completed
0.3 04-Nov-2013
 * Cookies are now Base64URL encoded, no more plain JSON visible
 * Working test suite, via http://personatestuser.org/
 * Require openssl
 * Get rid of custom HMAC implementation in favor of OpenSSL's (upgrade to
 * Cookies are now JSON encoded
   SHA256 as well)
0.2 29-Oct-2013
 * Bugfixes
0.1 28-Oct-2013
 * Internal-only cleanups
0.0.9 28-Oct-2013
 * Add AuthPersonaLogoutReturnTo configuration for post-logout redirecting
 * Add AuthPersonaLogoutURL configuration for logout handling
 * [bugfix] Persona Assertions expiry are in milliseconds
0.0.8 24-Oct-2013
 * Add AuthPersonaFakeBasicAuth On/Off directive
 * Fix per-server configuration merging
0.0.7 24-Oct-2013
 * Add AuthPersonaAuthoritative On/Off directive
0.0.6 22-Oct-2013
 * Fix per-directory configuration merging
 * Clear invalid/expired cookies
 * Moved AuthPersonaCookieSecure to per-directory configuration
 * Moved AuthPersonaCookieDomain to per-directory configuration
 * Moved AuthPersonaCookieName to per-directory configuration
 * Moved AuthPersonaCookieDuration to per-directory configuration
 * Moved AuthPersonaVerifierURL to per-directory configuration
 * Moved AuthPersonaLoginURL per-directory configuration
0.0.5
 * Added AuthPersonaCookieSecure for setting cookie https-only flag
 * Added AuthPersonaCookieDomain for setting cookie domain
0.0.4 18-Oct-2013
 * Added AuthPersonaCookieDuration for cookie expiration
 * Cookie now contains unforgable expiry time
0.0.3 17-Oct-2013
 * Replace yajl by json-c 
 * Added AuthPersonaLoginURL for configurable login location
0.0.2 15-Oct-2013
 * Initial release


src/mod_authnz_persona.c
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

#include "defines.h"
#include "cookie.h"
#include "verify.h"

#include <stdio.h>
#include <string.h>
#define APR_WANT_STRFUNC
#include <apr_want.h>
#include <apr_strings.h>
#include <apr_uuid.h>
#include <apr_tables.h>
#include <apr_random.h>

#include <httpd.h>
#include <http_config.h>
#include <http_core.h>
#include <http_log.h>
#include <http_protocol.h>
#include <http_request.h>       /* for ap_hook_(check_user_id | auth_checker) */
#include <apr_base64.h>
#if AP_MODULE_MAGIC_AT_LEAST(20080403, 1)
#include "ap_provider.h"
#include "mod_auth.h"
#endif

#include <curl/curl.h>
#include <curl/easy.h>
#include <assert.h>

#include "version.h"

/* apache module name */
module AP_MODULE_DECLARE_DATA authnz_persona_module;

const char *persona_server_secret_option(cmd_parms *, void *, const char *);
const char *persona_server_cookie_name(cmd_parms *, void *, const char *);
const char *persona_server_cookie_domain(cmd_parms *, void *, const char *);
const char *persona_server_cookie_duration(cmd_parms *, void *, const char *);
const char *persona_server_cookie_path(cmd_parms *, void *, const char *);
const char *persona_server_cookie_secure(cmd_parms *, void *, int);
const char *persona_authoritative(cmd_parms *, void *, int);
const char *persona_local_verify(cmd_parms *, void *, int);
const char *persona_server_verifier_url(cmd_parms *, void *, const char *);
const char *persona_server_login_url(cmd_parms *, void *, const char *);
const char *persona_server_logout_url(cmd_parms *, void *, const char *);
const char *persona_server_logout_returnto_url(cmd_parms *, void *,
                                               const char *);
const char *persona_fake_basic_auth(cmd_parms *, void *, int);
static apr_status_t persona_generate_secret(apr_pool_t *, server_rec *,
                                            persona_config_t *);

static int persona_authn_active(request_rec *r)
{
  return (strncmp("Persona", ap_auth_type(r), 8) == 0) ? 1 : 0;
}

static void fake_basic_auth(request_rec *r)
{
  char *basic = apr_pstrcat(r->pool, r->user, ":", "password", NULL);
  apr_size_t size = (apr_size_t) strlen(basic);
  char *base64 = apr_palloc(r->pool,
                            apr_base64_encode_len(size + 1) * sizeof(char));
  apr_base64_encode(base64, basic, size);
  apr_table_setn(r->headers_in, "Authorization",
                 apr_pstrcat(r->pool, "Basic ", base64, NULL));
  return;
}

static void set_cookie_from_config(persona_dir_config_t * dconf,
                                   Cookie cookie)
{
  cookie->expires = dconf->cookie_duration;
  cookie->domain = dconf->cookie_domain;
  cookie->secure = dconf->cookie_secure;
  cookie->path = dconf->cookie_path;
}

/**************************************************
 * Authentication phase
 *
 * - If AuthType != Persona, do nothing
 * - Handle POSTed assertions ("null" -> logout)
 * - If we have a cookie, set up user context
 **************************************************/
static int Auth_persona_check_cookie(request_rec *r)
{
  char *szCookieValue = NULL;
  const char *assertion = NULL;

  if (!persona_authn_active(r)) {
    return DECLINED;
  }

  persona_config_t *conf =
    ap_get_module_config(r->server->module_config, &authnz_persona_module);
  persona_dir_config_t *dconf =
    ap_get_module_config(r->per_dir_config, &authnz_persona_module);

  apr_table_set(r->err_headers_out, "X-Mod-Auth-Persona", VERSION);

  /* We take over all HTTP_UNAUTHORIZED pages */
  ap_custom_response(r, HTTP_UNAUTHORIZED, dconf->login_url);

  /* Assertions should only appear on POST requests */
  if (r->method_number == M_POST) {
    assertion = apr_table_get(r->headers_in, dconf->assertion_header);
  }

  // We'll trade you a valid assertion for a session cookie!
  // this is a programmatic XHR request
  if (assertion) {

     /* null assertion is a form of logout */
    if (!strncmp(assertion, "null", 5)) {
      /* XXX: Absctraction not quite right, creating a cookie structure here feels wrong */
      Cookie cookie = apr_pcalloc(r->pool, sizeof(*cookie));
      set_cookie_from_config(dconf, cookie);
      clearCookie(r, conf->secret, dconf->cookie_name, cookie);
      r->status = HTTP_OK;
      const char *status = "{\"status\": \"okay\"}";
      ap_set_content_type(r, "application/json");
      ap_rwrite(status, strlen(status), r);
      return DONE;
    }
  
    VerifyResult res;
    if (dconf->local_verify) {
      res = verify_assertion_local(r, assertion);
    }
    else {
      res = processAssertion(r, dconf->verifier_url, assertion);
    }

    if (!res->errorResponse) {
      assert(res->verifiedEmail);
      assert(res->identityIssuer);

      ap_log_rerror(APLOG_MARK, APLOG_INFO | APLOG_NOERRNO, 0, r, ERRTAG
                    "email '%s' verified, vouched for by issuer '%s'",
                    res->verifiedEmail, res->identityIssuer);

      Cookie cookie = apr_pcalloc(r->pool, sizeof(struct _Cookie));
      cookie->verifiedEmail = res->verifiedEmail;
      cookie->identityIssuer = res->identityIssuer;
      set_cookie_from_config(dconf, cookie);

      sendSignedCookie(r, conf->secret, dconf->cookie_name, cookie);

      r->user = apr_pstrdup(r->pool, res->verifiedEmail);

      /* XXX: At this point, we have authenticated the user, but we bail out too soon
       * XXX: from the processing. For this request completion, there is no r->user
       * XXX: However, this is the XHR request, so bail out before sending content out
       */
      return DONE;
    }
    else {
      ap_set_content_type(r, "application/json");
      ap_rwrite(res->errorResponse, strlen(res->errorResponse), r);
      apr_table_set(r->err_headers_out, "X-Persona-Error",
                    res->errorResponse);

      ap_log_rerror(APLOG_MARK, APLOG_INFO | APLOG_NOERRNO, 0, r, ERRTAG
                    "Assertion verification failed (%s)", res->errorResponse);

      // upon assertion verification failure we return JSON explaining why
      r->status = HTTP_INTERNAL_SERVER_ERROR;
      return DONE;
    }
  }

  // if there's a valid cookie, allow the user through
  szCookieValue = extractCookie(r, conf->secret, dconf->cookie_name);

  Cookie cookie = NULL;
  if (szCookieValue) {
    if ((cookie = validateCookie(r, conf->secret, szCookieValue))) {
      r->user = (char *) cookie->verifiedEmail;
      apr_table_setn(r->notes, PERSONA_ISSUER_NOTE, cookie->identityIssuer);
      apr_table_setn(r->subprocess_env, PERSONA_ENV_IDP,
                     cookie->identityIssuer);

      /* If requested, fake a Authorization: header */
      if (dconf->fake_basic_auth) {
        fake_basic_auth(r);
      }

      /* Logged-in user is visiting the logout url */
      if (dconf->logout_url
          && strncmp(dconf->logout_url, r->uri,
                     strlen(dconf->logout_url) + 1) == 0) {
        ap_log_rerror(APLOG_MARK, APLOG_DEBUG | APLOG_NOERRNO, 0, r,
                      ERRTAG "User '%s' logging out via '%s', sending to %s",
                      r->user, r->uri, dconf->logout_returnto_url);
        apr_table_setn(r->subprocess_env, PERSONA_ENV_LOGOUT_RETURNTO,
                       dconf->logout_returnto_url);
        set_cookie_from_config(dconf, cookie);
        clearCookie(r, conf->secret, dconf->cookie_name, cookie);
      }
      return OK;
    }
    else {                      /* cookie didn't validate */
      /* XXX: Absctraction not quite right, creating a cookie structure here feels wrong */
      cookie = apr_pcalloc(r->pool, sizeof(*cookie));
      set_cookie_from_config(dconf, cookie);
      clearCookie(r, conf->secret, dconf->cookie_name, cookie);
    }
  }

  return HTTP_UNAUTHORIZED;
}

#if !AP_MODULE_MAGIC_AT_LEAST(20080403, 1)

/**************************************************
 * Authorization phase (Apache 2.2)
 *
 * Requires authentication phase to run first.
 *
 * Handles Require persona-idp directives.
 **************************************************/
static int Auth_persona_check_auth(request_rec *r)
{
  const apr_array_header_t *reqs_arr = NULL;
  require_line *reqs = NULL;
  register int x;
  const char *szRequireLine;
  char *szRequire_cmd;

  if (!persona_authn_active(r)) {
    return DECLINED;
  }

  persona_dir_config_t *dconf =
    ap_get_module_config(r->per_dir_config, &authnz_persona_module);

  apr_table_set(r->err_headers_out, "X-Mod-Auth-Persona", VERSION);

  /* We take over all HTTP_UNAUTHORIZED pages */
  ap_custom_response(r, HTTP_UNAUTHORIZED, dconf->login_url);

  /* get require line */
  reqs_arr = ap_requires(r);
  reqs = reqs_arr ? (require_line *) reqs_arr->elts : NULL;

  /* decline if no require line found */
  if (!reqs_arr)
    return DECLINED;

  /* walk through the array to check each require command */
  for (x = 0; x < reqs_arr->nelts; x++) {

    if (!(reqs[x].method_mask & (AP_METHOD_BIT << r->method_number)))
      continue;

    /* get require line */
    szRequireLine = reqs[x].requirement;

    /* get the first word in require line */
    szRequire_cmd = ap_getword_white(r->pool, &szRequireLine);

    // persona-idp: check host part of user name
    if (!strncmp("persona-idp", szRequire_cmd, 12)) {
      char *reqIdp = ap_getword_conf(r->pool, &szRequireLine);
      const char *issuer = apr_table_get(r->notes, PERSONA_ISSUER_NOTE);
      if (!issuer || strncmp(issuer, reqIdp, strlen(reqIdp) + 1)) {
        ap_log_rerror(APLOG_MARK, APLOG_INFO | APLOG_NOERRNO, 0, r,
                      ERRTAG
                      "user '%s' is not authorized by idp:%s, but idp:%s instead",
                      r->user, reqIdp, (issuer ? issuer : "unknown"));

        char *error = apr_psprintf(r->pool,
                                   "user '%s' is not authenticated by IdP '%s' (but by '%s')",
                                   r->user, reqIdp,
                                   (issuer ? issuer : "unknown"));
        apr_table_setn(r->subprocess_env, "PERSONA_ERROR", error);
        return DECLINED;
      }

      ap_log_rerror(APLOG_MARK, APLOG_INFO | APLOG_NOERRNO, 0, r,
                    ERRTAG "user '%s' is authorized by idp:%s", issuer,
                    r->user);
      return OK;
    }
  }

  /* give others a chance */
  if (dconf->authoritative) {
    return HTTP_UNAUTHORIZED;
  }
  else {
    return DECLINED;
  }
}

#else

/**************************************************
 * Authorization phase (Apache 2.4)
 *
 * Handles Require persona-idp directives.
 *
 * When this is first called, the authentication context hasn't been setup
 * yet. Return AUTHZ_DENIED_NO_USER to force it to run, then this will be
 * called again, with the context setup.
 **************************************************/
static authz_status persona_idp_check_authorization(request_rec *r,
                                                    const char *require_args,
                                                    const void *parsed_require_args) {

  ap_log_rerror(APLOG_MARK, APLOG_ERR|APLOG_NOERRNO, 0, r, ERRTAG "Require persona-idp");
  if (!r->user)
    // this triggers running authn hook, which we need
    return AUTHZ_DENIED_NO_USER;

  char *reqIdp = ap_getword_white(r->pool, &require_args);
  const char *issuer = apr_table_get(r->notes, PERSONA_ISSUER_NOTE);
  return issuer && !strcmp(issuer, reqIdp) ? AUTHZ_GRANTED : AUTHZ_DENIED;
}

static const authz_provider authz_persona_idp_provider =
{
  &persona_idp_check_authorization,
  NULL,
};

#endif

static int Auth_persona_post_config(apr_pool_t * pconf, apr_pool_t * plog,
                                    apr_pool_t * ptemp, server_rec *s)
{
  server_rec *sp;
  persona_config_t *conf;

  for (sp = s; sp; sp = sp->next) {
    conf = ap_get_module_config(sp->module_config, &authnz_persona_module);
    if (!conf->secret->len) {
      persona_generate_secret(pconf, sp, conf);
      ap_log_error(APLOG_MARK, APLOG_INFO | APLOG_NOERRNO, 0, sp,
                   ERRTAG
                   "created a secret since none was configured for %s",
                   sp->server_hostname);
    }
  }

  ap_add_version_component(pconf, "mod_authn_persona/" VERSION);

  return OK;
}

/**************************************************
 * register module hooks
 **************************************************/

static void register_hooks(apr_pool_t *p)
{
  // these hooks are executed in order, first is first.
#if AP_MODULE_MAGIC_AT_LEAST(20080403, 1)
  ap_hook_check_authn(Auth_persona_check_cookie, NULL, NULL, APR_HOOK_FIRST,
                      AP_AUTH_INTERNAL_PER_CONF);
  ap_register_auth_provider(p, AUTHZ_PROVIDER_GROUP, "persona-idp",
                            AUTHZ_PROVIDER_VERSION, &authz_persona_idp_provider,
                            AP_AUTH_INTERNAL_PER_CONF);
#else
  ap_hook_check_user_id(Auth_persona_check_cookie, NULL, NULL, APR_HOOK_FIRST);
  ap_hook_auth_checker(Auth_persona_check_auth, NULL, NULL, APR_HOOK_FIRST);
#endif
  ap_hook_post_config(Auth_persona_post_config, NULL, NULL, APR_HOOK_MIDDLE);
}

static apr_status_t persona_generate_secret(apr_pool_t * p, server_rec *s,
                                            persona_config_t * conf)
{
  unsigned char *secret = apr_pcalloc(p, conf->secret_size);
  apr_status_t status;

  status = apr_generate_random_bytes(secret, conf->secret_size);

  if (APR_SUCCESS == status) {
    /* Truncate to right length */
    secret[conf->secret_size] = 0;
    conf->secret->data = (char *) secret;
    conf->secret->len = conf->secret_size;
  }

  return status;
}

static void *persona_create_dir_config(apr_pool_t * p, char *path)
{
  persona_dir_config_t *dconf = apr_palloc(p, sizeof(*dconf));

  dconf->verifier_url = PERSONA_DEFAULT_VERIFIER_URL;
  dconf->verifier_url_set = 0;
  dconf->login_url = PERSONA_LOGIN_URL;
  dconf->login_url_set = 0;
  dconf->logout_url = NULL;
  dconf->logout_url_set = 0;
  dconf->logout_returnto_url = "/";
  dconf->logout_returnto_url_set = 0;
  dconf->cookie_secure = 0;
  dconf->cookie_secure_set = 0;
  dconf->authoritative = 0;
  dconf->authoritative_set = 0;
  dconf->cookie_path = "/";
  dconf->cookie_path_set = 0;
  dconf->cookie_name = PERSONA_COOKIE_NAME;
  dconf->cookie_name_set = 0;
  dconf->cookie_domain = NULL;
  dconf->cookie_domain_set = 0;
  dconf->cookie_duration = PERSONA_COOKIE_DURATION;
  dconf->cookie_duration_set = 0;
  dconf->assertion_header = PERSONA_ASSERTION_HEADER;
  dconf->assertion_header_set = 0;
  dconf->local_verify = 0;
  dconf->local_verify_set = 0;
  return dconf;
}

static void *persona_create_svr_config(apr_pool_t * p, server_rec *s)
{
  persona_config_t *conf = apr_palloc(p, sizeof(*conf));

  conf->secret = apr_pcalloc(p, sizeof(buffer_t));
  conf->secret_size = PERSONA_SECRET_SIZE;


  return conf;
}

const char *persona_server_secret_option(cmd_parms *cmd, void *cfg,
                                         const char *arg)
{
  server_rec *s = cmd->server;
  persona_config_t *conf =
    ap_get_module_config(s->module_config, &authnz_persona_module);
  conf->secret->len = strlen(arg);
  conf->secret->data = apr_palloc(cmd->pool, conf->secret->len);
  strncpy(conf->secret->data, arg, conf->secret->len);
  return NULL;
}

const char *persona_server_cookie_duration(cmd_parms *cmd, void *cfg,
                                           const char *arg)
{
  persona_dir_config_t *dconf = cfg;
  dconf->cookie_duration = atoi(arg);
  dconf->cookie_duration_set = 1;
  return NULL;
}
const char *persona_server_cookie_path(cmd_parms *cmd, void *cfg,
                                       const char *arg)
{
  persona_dir_config_t *dconf = cfg;
  dconf->cookie_path = apr_pstrdup(cmd->pool, arg);
  dconf->cookie_path_set = 1;
  return NULL;
}
const char *persona_server_cookie_name(cmd_parms *cmd, void *cfg,
                                       const char *arg)
{
  persona_dir_config_t *dconf = cfg;
  dconf->cookie_name = apr_pstrdup(cmd->pool, arg);
  dconf->cookie_name_set = 1;
  return NULL;
}

const char *persona_server_cookie_secure(cmd_parms *cmd, void *cfg, int flag)
{
  persona_dir_config_t *dconf = cfg;
  dconf->cookie_secure = flag;
  dconf->cookie_secure_set = 1;
  return NULL;
}

const char *persona_local_verify(cmd_parms *cmd, void *cfg, int flag)
{
  persona_dir_config_t *dconf = cfg;
  dconf->local_verify = flag;
  dconf->local_verify_set = 1;
  return NULL;
}

const char *persona_authoritative(cmd_parms *cmd, void *cfg, int flag)
{
  persona_dir_config_t *dconf = cfg;
  dconf->authoritative = flag;
  dconf->authoritative_set = 1;
  return NULL;
}

const char *persona_fake_basic_auth(cmd_parms *cmd, void *cfg, int flag)
{
  persona_dir_config_t *dconf = cfg;
  dconf->fake_basic_auth = flag;
  dconf->fake_basic_auth_set = 1;
  return NULL;
}

const char *persona_server_cookie_domain(cmd_parms *cmd, void *cfg,
                                         const char *arg)
{
  persona_dir_config_t *dconf = cfg;
  dconf->cookie_domain = apr_pstrdup(cmd->pool, arg);
  dconf->cookie_domain_set = 1;
  return NULL;
}

const char *persona_server_verifier_url(cmd_parms *cmd, void *cfg,
                                        const char *arg)
{
  persona_dir_config_t *dconf = cfg;
  dconf->verifier_url = apr_pstrdup(cmd->pool, arg);
  dconf->verifier_url_set = 1;
  return NULL;
}

const char *persona_server_login_url(cmd_parms *cmd, void *cfg,
                                     const char *arg)
{
  persona_dir_config_t *dconf = cfg;
  dconf->login_url = apr_pstrdup(cmd->pool, arg);
  dconf->login_url_set = 1;
  return NULL;
}

const char *persona_server_logout_url(cmd_parms *cmd, void *cfg,
                                      const char *arg)
{
  persona_dir_config_t *dconf = cfg;
  dconf->logout_url = apr_pstrdup(cmd->pool, arg);
  dconf->logout_url_set = 1;
  return NULL;
}

const char *persona_server_logout_returnto_url(cmd_parms *cmd, void *cfg,
                                               const char *arg)
{
  persona_dir_config_t *dconf = cfg;
  dconf->logout_returnto_url = apr_pstrdup(cmd->pool, arg);
  dconf->logout_returnto_url_set = 1;
  return NULL;
}

/* If the current config is set, use it, otherwise, use the parent's */
#define persona_merge_parent(name, merged, parent, child) \
  merged->name = child->name ## _set ? child->name : parent->name; \
  merged->name ## _set = child->name ## _set ? child->name ## _set : parent->name ## _set

static void *persona_merge_dir_config(apr_pool_t * p, void *parent_conf,
                                      void *child_conf)
{
  persona_dir_config_t *parent = (persona_dir_config_t *) parent_conf;
  persona_dir_config_t *child = (persona_dir_config_t *) child_conf;
  persona_dir_config_t *merged = apr_pcalloc(p, sizeof(*merged));

  persona_merge_parent(cookie_path, merged, parent, child);
  persona_merge_parent(cookie_name, merged, parent, child);
  persona_merge_parent(cookie_domain, merged, parent, child);
  persona_merge_parent(cookie_duration, merged, parent, child);
  persona_merge_parent(cookie_secure, merged, parent, child);
  persona_merge_parent(authoritative, merged, parent, child);
  persona_merge_parent(login_url, merged, parent, child);
  persona_merge_parent(logout_url, merged, parent, child);
  persona_merge_parent(logout_returnto_url, merged, parent, child);
  persona_merge_parent(verifier_url, merged, parent, child);
  persona_merge_parent(assertion_header, merged, parent, child);
  persona_merge_parent(fake_basic_auth, merged, parent, child);
  persona_merge_parent(local_verify, merged, parent, child);

  return merged;
}

static void *persona_merge_svr_config(apr_pool_t * p, void *parent_conf,
                                      void *child_conf)
{
  persona_config_t *parent = (persona_config_t *) parent_conf;
  persona_config_t *child = (persona_config_t *) child_conf;
  persona_config_t *merged = apr_pcalloc(p, sizeof(*merged));

  if (child->secret->len) {
    merged->secret_size = child->secret_size;
    merged->secret = child->secret;
  }
  else {
    merged->secret_size = parent->secret_size;
    merged->secret = parent->secret;
  }

  return merged;
}

static const command_rec Auth_persona_options[] = {
  AP_INIT_TAKE1("AuthPersonaServerSecret", persona_server_secret_option,
                NULL, RSRC_CONF, "Server secret to use for cookie signing"),
  AP_INIT_TAKE1("AuthPersonaCookieName", persona_server_cookie_name,
                NULL, RSRC_CONF | OR_AUTHCFG, "Name of the Persona Cookie"),
  AP_INIT_TAKE1("AuthPersonaCookieDomain", persona_server_cookie_domain,
                NULL, RSRC_CONF | OR_AUTHCFG,
                "Domain for the Persona Cookie"),
  AP_INIT_TAKE1("AuthPersonaCookieDuration", persona_server_cookie_duration,
                NULL, RSRC_CONF | OR_AUTHCFG,
                "Duration of the Persona Cookie"),
  AP_INIT_TAKE1("AuthPersonaCookiePath", persona_server_cookie_path,
               NULL, RSRC_CONF | OR_AUTHCFG, "Persona Cookie Path"),		
  AP_INIT_FLAG("AuthPersonaCookieSecure", persona_server_cookie_secure,
               NULL, RSRC_CONF | OR_AUTHCFG, "HTTPS only Persona Cookie"),
  AP_INIT_FLAG("AuthPersonaAuthoritative", persona_authoritative,
               NULL, RSRC_CONF | OR_AUTHCFG, "HTTPS only Persona Cookie"),
  AP_INIT_FLAG("AuthPersonaLocalVerify", persona_local_verify,
               NULL, RSRC_CONF | OR_AUTHCFG,
               "Perform local assertion verification"),
  AP_INIT_TAKE1("AuthPersonaVerifierURL", persona_server_verifier_url,
                NULL, RSRC_CONF | OR_AUTHCFG,
                "URL to a Persona Verfier service"),
  AP_INIT_TAKE1("AuthPersonaLoginURL", persona_server_login_url,
                NULL, RSRC_CONF | OR_AUTHCFG, "URL to a Persona login page"),
  AP_INIT_TAKE1("AuthPersonaLogoutURL", persona_server_logout_url,
                NULL, RSRC_CONF | OR_AUTHCFG, "URL to a Persona logout page"),
  AP_INIT_TAKE1("AuthPersonaLogoutReturnTo",
                persona_server_logout_returnto_url,
                NULL, RSRC_CONF | OR_AUTHCFG,
                "URL to redirect to after logging out"),
  AP_INIT_FLAG("AuthPersonaFakeBasicAuth", persona_fake_basic_auth,
               NULL, RSRC_CONF | OR_AUTHCFG,
               "Should we fake basic authentication?"),
  {NULL}
};

/* apache module structure */
module AP_MODULE_DECLARE_DATA authnz_persona_module = {
  STANDARD20_MODULE_STUFF,
  persona_create_dir_config,    /* dir config creator */
  persona_merge_dir_config,     /* dir merger --- default is to override */
  persona_create_svr_config,    /* server config creator */
  persona_merge_svr_config,     /* merge server config */
  Auth_persona_options,         /* command apr_table_t */
  register_hooks                /* register hooks */
};
